# Data Analytics Program
## Python Fundamentals for Data Analysts
## 4.4: Data Wrangling & Subsetting
## Alphy Jose

# Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Importing datasets

In [2]:
# Craeting string of project folder path
path = r'C:\Users\augus\Documents\01-2026 Instacart Basket Analysis' 

In [3]:
# importing orders.csv data set
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [14]:
# Importing products.csv data set
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [15]:
# Importing departments.csv data set
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

## Data Wrangling Examples

In [5]:
# Dropping eval_set column from the dataframe
df_ords = df_ords.drop(columns = ['eval_set'])

In [7]:
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [39]:
# Finding missing values
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [9]:
# Renaming column
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [10]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [11]:
# Changing a varaibles's data type
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [12]:
df_ords['order_id'].dtype

dtype('O')

In [17]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [21]:
# transposing department data
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [22]:
# create new dataframe for transposed department data
df_dep_t = df_dep.T

In [20]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [23]:
# adding index
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [24]:
#create a new header
new_header = df_dep_t.iloc[0]

In [25]:
new_header

0    department
Name: department_id, dtype: object

In [26]:
#create a new dataframe that only copies over rows beyond the first row.
df_dep_t_new = df_dep_t[1:]

In [27]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [28]:
# sets the header row as the df header
df_dep_t_new.columns = new_header

In [29]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# Data Dictionary

In [30]:
# creating data dictionary
data_dict = df_dep_t_new.to_dict('index')

In [31]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [32]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [33]:
print(data_dict.get('19'))

{'department': 'snacks'}


# Subsetting

In [34]:
# Creating subset of products in snacks department
df_snacks =  df_prods[df_prods['department_id']==19]

In [35]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# Task

## Step 2 :Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [64]:
# Changing the datatype of user_id from integer to string beacuse it doesn't need to be a number
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [37]:
df_ords['user_id'].dtype

dtype('O')

## Step 3:Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

#### The column order_dow has alreay been renamed to orders_day_of_week. I couldnt find any other variable in df_ords dataframe that shpuld be renamed

## Step 4: Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [65]:
# checking the frequency of the column order_hour_of_day to find the busiest hour for placing orders
df_ords['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

#### The busiest hour of the day for placing orders is 10:00 AM 

## Step 5 :Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary

In [66]:
# printing the meaning behind value 4 in 'department_id' column using data dictionary
print(data_dict.get('4'))

{'department': 'produce'}


## Step 6 :The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [42]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [43]:
# Creating subset of breakfast items from df_prods
df_breakfast =  df_prods[df_prods['department_id']==14]

In [44]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

## Step 7 : The sales team in your client’s organization wants to know more about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood.

In [45]:
# Creating subset of products for dinner parties
df_dinnerparty =  df_prods[df_prods['department_id'].isin([5,20,7,12])]

In [46]:
df_dinnerparty.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

## Step 8 : How many rows does the last dataframe you created have?

In [67]:
# finding the number of rows and columns in dinnerparty dataframe using shape function
df_dinnerparty.shape

(7650, 5)

#### There are 7650 rows in df_dinnerparty dataframe

## Step 9 :Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [68]:
# Creating a dataframe with details of customer with user_id 1
df_user1 = df_ords[df_ords['user_id'] == '1']

In [52]:
df_user1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

## Step 10 :You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [69]:
# Using describe function to fetch basic statistics about user_1
df_user1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

In [54]:
df_user1['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
8     3
7     3
12    1
15    1
9     1
14    1
16    1
Name: count, dtype: int64

In [57]:
df_user1['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
4    4
1    3
2    2
3    2
Name: count, dtype: int64

#### User1 has made a total of 11 orders with most orders made on wednesdays and sundays.These orders are placed between 7:00 and 16:00, with the majority of them placed around 7.00 and 8.00. On average, this user places an order every 19 days.

## Step 12 :Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [62]:
# export df_ords dataframe
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'),index=False)

## Step 13:Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder 

In [63]:
# export df_dep_t_new dataframe
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'),index=False)
